src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,761 @@
+
+	op->sgl.sg_nr     = 1;
+	op->sgl.sg_nr_out = 0;
+	op->sgl.sg_iovs   = &op->iov;
+
+	rc = dfs_read(hdl->dfs, op->obj, &op->sgl, 0 /* offset */, &op->size, &op->ev);
+	if (rc) {
+		D_ERROR("Could not start async read on '%s': %s (rc=%d)", op->path, strerror(rc),
+			rc);
+		goto out;
+	}
+
+	return 0;
+
+out:
+	PyBuffer_Release(&op->buf_view);
+
+	int rc2 = daos_event_fini(&op->ev);
+	if (rc2) {
+		D_ERROR("Could not finalize event: %s (rc=%d)", d_errstr(rc2), rc2);
+	}
+
+	if (op->obj != NULL) {
+		rc2 = dfs_release(op->obj);
+		if (rc2) {
+			D_ERROR("Could not release obj handler of '%s': %s (rc=%d)", op->path,
+				strerror(rc2), rc2);
+		}
+		op->obj = NULL;
+	}
+
+	return rc;
+}
+
+static int
+complete_read_op(struct dfs_handle *hdl, struct io_op *op)
+{
+	assert(op != NULL);
+
+	D_DEBUG(DB_ANY, "READ of %zu bytes from '%s' completed with  status: %s (rc = %d)",
+		op->size, op->path, d_errstr(op->ev.ev_error), op->ev.ev_error);
+
+	int rc = dfs_release(op->obj);
+	if (rc) {
+		D_WARN("Could not release object handler %s: %s (rc=%d)", op->path, strerror(rc),
+		       rc);
+	}
+	op->obj = NULL;
+	op->err = op->ev.ev_error;
+
+	if (op->err == 0 && op->size != op->buf_view.len) {
+		op->err = EIO;
+	}
+
+	rc = daos_event_fini(&op->ev);
+	if (rc) {
+		D_WARN("Could not finalize event handler of '%s': %s (rc=%d)", op->path,
+		       d_errstr(rc), rc);
+	}
+
+	PyBuffer_Release(&op->buf_view);
+	return 0;
+}
+
+/*
+
+ */
+static PyObject *
+__shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+{
+	int                rc        = 0;
+	int                completed = 0;
+	PyObject          *items     = NULL;
+	Py_ssize_t         nr        = 0;
+	struct io_op      *ops       = NULL;
+	struct dfs_handle *hdl       = NULL;
+	daos_event_t      *evp[EQ_POLL_BATCH_SIZE];
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LO", &hdl, &items);
+	assert(hdl->dfs != NULL);
+
+	nr = PyList_Size(items);
+	if (nr <= 0) {
+		return PyLong_FromLong(EINVAL);
+	}
+
+	D_ALLOC_ARRAY(ops, nr);
+	if (ops == NULL) {
+		return PyLong_FromLong(ENOMEM);
+	}
+
+	/* For optimal usage of transport layer, we try to enqueue all items in the bath
+	   leaving the throttling up to transport level.
+	   It can be manually adjusted by D_QUOTA_RPC environment variable.
+	 */
+	for (Py_ssize_t i = 0; i < nr; ++i) {
+		PyObject *item = PyList_GetItem(items, i);
+		if (item == NULL) {
+			D_ERROR("Unexpected NULL entry in the batch read list");
+			rc = EINVAL;
+
+			nr = i;
+			break;
+		}
+
+		rc = start_read_op(hdl, item, &ops[i]);
+		if (rc) {
+			nr = i;
+			break;
+		}
+	}
+
+	do {
+		int eq_rc = daos_eq_poll(hdl->eq, 0, DAOS_EQ_NOWAIT, nr, evp);
+		if (eq_rc < 0) {
+			D_ERROR("Could not poll event queue: %s (rc=%d)", d_errstr(eq_rc), eq_rc);
+			rc = daos_der2errno(eq_rc);
+			break;
+		}
+
+		for (int i = 0; i < eq_rc; ++i) {
+			struct io_op *op    = container_of(evp[i], struct io_op, ev);
+			int           op_rc = complete_read_op(hdl, op);
```

Comment:
complete_read_op() sets op->err but i missed where this err is checked to indicate a read error has occurred (if it did)?

Reply:
unresolving as it sounds like we can still overwrite the error on a subsequent iteration? and the return of that function would still be SUCCESS?

Reply:
the function returns `rc` and this loop it can be overwritten only if it's 0 , so only first encountered error code will be set 😬

Reply:
@mchaarawi ^ I'm not sure why it does not show reply in the comments section 😬

Reply:
you are correct, my bad